docs/spec/fastsync/fastsync.md

```diff
@@ -0,0 +1,1101 @@
+# Fastsync
+
+<!--
+> Rough outline of what the component is doing and why. 2-3 paragraphs 
+---->
+
+This document is the English specification of the *Fastsync*
+protocol. It consists of the following parts:
+
+- [Part I](#part-i---outside-view): Informal introduction; high-level (sequential) specification of the problem
+  addressed by *Fastsync*
+  
+- [Part II](#part-ii---protocol-view): Protocol view, 
+    - temporal logic specifications, 
+	- description of Fastsync V2 (the protocol underlying the current Golang
+      implementation)
+	- analysis of Fastsync V2 that highlights several issues that
+      prevent achieving some of the desired fault-tolerance properties
+
+- [Part III](#part-iii---suggestions-for-an-improved-fastsync-implementation): some suggestions on how to address the issues in the future
+  
+
+# Part I - Outside view
+
+## Context of this document
+
+<!--
+> mention other components and or specifications that are relevant for this
+spec. Possible interactions, possible use cases, etc. 
+---->
+<!--
+> should give the reader the understanding in what environment this component
+will be used. 
+---->
+
+Fastsync is a protocol that is used by a full node to catch-up to the
+current state of a Tendermint blockchain. Its typical use case is a
+full node that was disconnected from the system for some time. The
+recovering full node locally has a copy of a prefix of the blockchain,
+and the corresponding application state that is slightly outdated. It
+then queries its peers for the blocks that were decided on by the
+Tendermint blockchain during the period the full node was
+disconnected. After receiving these blocks, it executes the
+transactions in the blocks in order to catch-up to the current height
+of the blockchain and the corresponding application state.
+
+In practice it is sufficient to catch-up only close to the current
+height: The Tendermint consensus reactor implements a similar
+functionality and can synchronize a node that is approximately 100
```

Comment:
I don't remember seeing code that does this in consensus reactor. @ebuchman @melekes Can someone confirm this? Ideally we should also add a reference to the corresponding code here.

Reply:
Consensus reactor can catch up, but it's very slow comparing to fast sync & state sync.

Reply:
Ethan changed it to 10 already.